File: src/lib/ai/providers/ocrProvider.ts
import { opencodeText } from "./opencodeProvider";

const OCR_BASE_URL = "https://jebin2-ocr.hf.space";

export async function opencodeImage(
  imageBase64: string,
  mimeType: string,
  text: string,
  system: string,
): Promise<string> {
  const blob = new Blob([Buffer.from(imageBase64, "base64")], { type: mimeType });
  const form = new FormData();
  form.append("image", blob, "receipt.jpg");
  form.append("hide_from_ui", "true");

  const uploadRes = await fetch(`${OCR_BASE_URL}/api/tasks/upload`, {
    method: "POST",
    body: form,
  });
  if (!uploadRes.ok) throw new Error(`OCR upload failed: ${uploadRes.status}`);
  const { id: taskId } = await uploadRes.json() as { id: string };
  if (!taskId) throw new Error("OCR upload returned no task ID");

  let ocrText = "";
  const deadline = Date.now() + 90_000;
  while (Date.now() < deadline) {
    await new Promise((r) => setTimeout(r, 2000));
    const pollRes = await fetch(`${OCR_BASE_URL}/api/tasks/${taskId}`);
    if (!pollRes.ok) throw new Error(`OCR poll failed: ${pollRes.status}`);
    const task = await pollRes.json() as { status: string; result?: string; error?: string };
    if (task.status === "completed") {
      try {
        const parsed = JSON.parse(task.result ?? "{}") as { text?: string };
        ocrText = parsed.text ?? "";
      } catch {
        ocrText = task.result ?? "";
      }
      break;
    }
    if (task.status === "failed") throw new Error(`OCR task failed: ${task.error}`);
  }
  if (!ocrText) throw new Error("OCR returned empty text");

  const combined = [text, "---", "Text extracted from image:", ocrText].filter(Boolean).join("\n");
  return opencodeText(combined, system);
}

File: backend/app/ai/providers/ocr_provider.py
"""OCR + OpenCode image provider — port of src/lib/ai/providers/ocrProvider.ts.
Extracts text from the image via the HF OCR space, then answers via opencode_text."""
import asyncio
import base64
import json
import time

import httpx

from app.ai.providers.opencode_provider import auth_headers, opencode_text
from app.config import settings


def _base_url() -> str:
    return (settings.ocr_api_url or "https://jebin2-ocr.hf.space").rstrip("/")


async def opencode_image(image_base64: str, mime_type: str, text: str, system: str) -> str:
    base = _base_url()
    async with httpx.AsyncClient(timeout=30.0, headers=auth_headers()) as client:
        upload = await client.post(
            f"{base}/api/tasks/upload",
            files={"image": ("receipt.jpg", base64.b64decode(image_base64), mime_type)},
            data={"hide_from_ui": "true"},
        )
        if upload.status_code >= 400:
            raise RuntimeError(f"OCR upload failed: {upload.status_code}")
        task_id = upload.json().get("id")
        if not task_id:
            raise RuntimeError("OCR upload returned no task ID")

        ocr_text = ""
        deadline = time.monotonic() + 90.0
        while time.monotonic() < deadline:
            await asyncio.sleep(2.0)
            poll = await client.get(f"{base}/api/tasks/{task_id}")
            if poll.status_code >= 400:
                raise RuntimeError(f"OCR poll failed: {poll.status_code}")
            task = poll.json()
            if task.get("status") == "completed":
                try:
                    parsed = json.loads(task.get("result") or "{}")
                    ocr_text = parsed.get("text") or ""
                except (json.JSONDecodeError, TypeError):
                    ocr_text = task.get("result") or ""
                break
            if task.get("status") == "failed":
                raise RuntimeError(f"OCR task failed: {task.get('error')}")

    if not ocr_text:
        raise RuntimeError("OCR returned empty text")

    combined = "\n".join(p for p in [text, "---", "Text extracted from image:", ocr_text] if p)
    return await opencode_text(combined, system)
